browser/src/Services/Notifications/NotificationsView.tsx

```diff
@@ -181,42 +181,67 @@ const NotificationHeader = styled.header`
     border-bottom: 1px solid ${p => p.theme["toolTip.border"]};
     padding: 0.5rem;
 `
+interface IViewState {
+    expired: boolean
+}
+
+export class NotificationView extends React.PureComponent<INotification, IViewState> {
+    public state = {
+        expired: false,
+    }
 
-export class NotificationView extends React.PureComponent<INotification, {}> {
+    private _timer: any // NodeJS.Timer should work but there's a type conflict
+    private _lifetime = 20000
     private iconDictionary = {
         error: "times-circle",
         warn: "exclamation-triangle",
         info: "info-circle",
     }
 
+    public componentDidMount() {
+        if (this.props.level !== "error") {
```

Comment:
In terms of implementation - how about we add an `expirationTime` property to the `INotification`, and add an `Epic` to handle the timeout behavior? I'd like for a consumer of our `Notifications` API to be able to set an expiration time.

Ideally, the UI wouldn't need to know or care about this - it could be handled in our store layer.

Reply:
More specifically:

1) We could add a `setExpiration` method on the `Notification` API object here: https://github.com/onivim/oni/blob/9d685a98ccd9485678144be4eff2562bd3cf4352/browser/src/Services/Notifications/Notification.ts#L16
2) We could add an `expirationTime` property in the store here: 
https://github.com/onivim/oni/blob/9d685a98ccd9485678144be4eff2562bd3cf4352/browser/src/Services/Notifications/NotificationStore.ts#L24

And then we could add an `Epic` on the store that looks something like this:

```
const hideNotificationAfterExpirationEpic: Epic<NotificationAction, INotificationState> => (action$, store) =>
    action$.ofType("SHOW_NOTIFICATION")
    .filter((action) => action.expirationTime !== -1)
    .delay((action) => action.expirationTime)
    .map((action) => ({ type: "HIDE_NOTIFICATION", id: action.id }))
```

At least something like that. The cool thing is, the UI doesn't need to worry about adding timers / handling lifecycle events, and we get the benefit of being able to specify this in our API surface as opposed to having it baked into our UI code.

Reply:
@bryphe sounds cool when I was adding it it definitely seemed like something that should be configurable given that I arbitrarily chose the number.

Thanks for the snippets that's really helpful as I'm not as familiar with Epics as I'd like and I've seen them around I'll have a tinker locally and push that when I get it working

Reply:
> Thanks for the snippets that's really helpful as I'm not as familiar with Epics as I'd like and I've seen them around I'll have a tinker locally and push that when I get it working

Awesome, thanks @Akin909 ! Let me know if you need any help.

The `ExplorerStore` uses just a couple of epics, hopefully those are good examples:
https://github.com/onivim/oni/blob/30c186aec1496ac9a8f6aa2af93a2b9fc98e0646/browser/src/Services/Explorer/ExplorerStore.ts#L191

(There's also some initialization required to 'wire up' the epics - example here: https://github.com/onivim/oni/blob/30c186aec1496ac9a8f6aa2af93a2b9fc98e0646/browser/src/Services/Explorer/ExplorerStore.ts#L228)

Reply:
@bryphe what looked like a slam dunk given your detailed snippet, has become a bit of a head scratcher. This is what i've got
```ts
const hideNotificationAfterExpirationEpic: Epic<NotificationAction, INotificationsState> = (
    action$,
    store,
) => {
    let currentExpTime: number
    return action$
        .ofType("SHOW_NOTIFICATION")
        .filter((action: IShowNotification) => action.expirationTime === -1)
        .map((action: IShowNotification) => (currentExpTime = action.expirationTime))
        .delay(currentExpTime)
        .map((action: IShowNotification) => ({ type: "HIDE_NOTIFICATION", id: action.id }))
}
```
except the map function at then end there gets passed a number by the delay function, not the action and I've had to do this not ideal thing of saving the exp time to an external var because the delay function doesn't get passed the action passed into it.

Had a brief look at the docs and couldn't find the correct alternative methods easily will have another look in a bit but thought I'd pick your brain first

Reply:
Ah interesting, I thought `delay` might have an overload like:
```
 .delay((action) => action.expirationTime)
```

But I guess that's not the case? If not, I'm okay with how you have it set up right now. It is a side effect but at least it's contained

Reply:
I'm quite interested in getting it to work with the `Epic` would be very cool will definitely spend a bit more time on it but can merge as is and I refactor it to the Epic in another PR or do it here?

Reply:
I'd prefer the `Epic` implementation for this PR, but let me know if there are any blockers.